src/EditorFeatures/Core/Implementation/IntelliSense/Completion2/RoslynCompletionItemSource.cs

```diff
@@ -0,0 +1,201 @@
+    [Name("C# and Visual Basic Completion Item Source")]
+    [ContentType(ContentTypeNames.VisualBasicContentType)]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    internal class RoslynCompletionItemSource : IAsyncCompletionItemSource
+    {
+        private ImmutableArray<char> CommitChars => ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+        private const string RoslynItem = nameof(RoslynItem);
+        private const string TriggerSnapshot = nameof(TriggerSnapshot);
+
+        public async Task<EditorCompletion.CompletionContext> GetCompletionContextAsync(
+            EditorCompletion.CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation,
+            CancellationToken cancellationToken)
+        {
+            var snapshot = triggerLocation.Snapshot;
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                // TODO: return default;
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty, new SnapshotSpan(triggerLocation.Snapshot, new Span(0, 0)));
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(document, triggerLocation.Position).ConfigureAwait(false);
+            if (completionList == null)
+            {
+                // TODO: return default;
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty, new SnapshotSpan(triggerLocation.Snapshot, new Span(0, 0)));
+            }
+
+            var text = await document.GetTextAsync().ConfigureAwait(false);
+            var applicableSpan = completionService.GetDefaultCompletionListSpan(text, triggerLocation.Position).ToSnapshotSpan(triggerLocation.Snapshot);
+
+            var imageIdService = document.Project.Solution.Workspace.Services.GetService<IImageIdService>();
+
+            Dictionary<string, CompletionFilter> filterCache = new Dictionary<string, CompletionFilter>();
+
+            var service = GetCompletionService(triggerLocation.Snapshot.TextBuffer.CurrentSnapshot) as CompletionServiceWithProviders;
+
+            var items = completionList.Items.SelectAsArray(roslynItem =>
+            {
+                var needsCustomCommit = service.GetProvider(roslynItem) is ICustomCommitCompletionProvider;
+
+                var item = Convert(roslynItem, imageIdService, completionService, filterCache, needsCustomCommit);
+                item.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                return item;
+            });
+
+            return new EditorCompletion.CompletionContext(
```

Comment:
would it be good for perf to also send a list of filters used?
Currently, in platform code, I go through all items received from all sources:
```
            var availableFilters = initialCompletionItems
                .SelectMany(n => n.Filters)
                .Distinct()
```
If you sent me filters, this would be 
```
            var availableFilters = initialCompletionFilters
                .SelectMany(n => n.Filters)
                .Distinct()
```
(the first `SelectMany` enumeration would be shorter)